### PR TITLE
[Merged by Bors] - Logging via the HTTP API

### PR DESCRIPTION
**Pull request by AgeManning**
This PR adds the ability to read the Lighthouse logs from the HTTP API for both the BN and the VC. 

This is done in such a way to as minimize any kind of performance hit by adding this feature.

The current design creates a tokio broadcast channel and mixes is into a form of slog drain that combines with our main global logger drain, only if the http api is enabled. 

The drain gets the logs, checks the log level and drops them if they are below INFO. If they are INFO or higher, it sends them via a broadcast channel only if there are users subscribed to the HTTP API channel. If not, it drops the logs. 

If there are more than one subscriber, the channel clones the log records and converts them to json in their independent HTTP API tasks.

### CODE REVIEW

**Comment by paulhauner** [comment]
It looks like this one didn't make it, sorry Age :(

File: book/src/api-lighthouse.md [hunk 1]
```diff
@@ -677,3 +677,32 @@ Caveats:
...
+Example:
+
+```bash
+curl -N "http://localhost:5052/lighthouse/logs"
+```
+
+Should provide an output that emits log events as they occur:
+```json
+{ 
+"data": {
```
Comment (amakrushin) [inline book/src/api-lighthouse.md:698]: in other places response format is `{{ ... }}`, but here - `{ "data": { ... }}`

**Review by amakrushin — COMMENTED** [review]
not sure if I am right or not.

**Comment by AgeManning** [comment]
@atokama - Thanks for the review. I'll change it to be consistent :).

File: book/src/api-lighthouse.md (hunk 1 above)
Comment (AgeManning) [reply book/src/api-lighthouse.md:698]: The majority of endpoints used seem to include data. There are a few that don't which return lists

**Comment by AgeManning** [comment]
This is starting to get integrated into the UI. Can I get a review on this at some stage @paulhauner @michaelsproul

**Comment by paulhauner** [comment]
I'll start reviewing this today or tomorrow so we can get it into v4.2.0.

**Comment by AgeManning** [comment]
During a small testnet with a few validators, the BN appears to dump out quite a few INFO logs. So much that the UI struggles to keep up. 

I'm hesitant to add a CLI flag that specifies the debug level for the SSE logs because we are always going to need at least INFO when displaying things. It might be handy for devs if they want debug logs coming from the HTTP API however. 

Curious about others thoughts on whether we should add a CLI flag to specify the filter level. 

For the testnet, we could in the meantime just change this line to whatever level we want: https://github.com/sigp/lighthouse/pull/4074/files#diff-5ed3cabe06cbfef735112f6092f9f5e964a618e4e8e3617ee182e9415820efb8R13

File: book/src/api-vc-endpoints.md [hunk 2]
```diff
@@ -578,3 +578,33 @@ The following fields may be omitted or nullified to obtain default values:
...
+Provides a subscription to receive logs as Server Side Events. Currently the
+logs emitted are INFO level or higher.
+
+### HTTP Specification
+
+| Property          | Specification                              |
+|-------------------|--------------------------------------------|
+| Path              | `/lighthouse/logs`                    |
+| Method            | GET                                        |
+| Required Headers  | [`Authorization`](./api-vc-auth-header.md) |
```
Comment (paulhauner) [inline book/src/api-vc-endpoints.md]:
```suggestion
| Required Headers  | None |
```

I understand that auth isn't required for this endpoint.

File: common/logging/src/sse_logging_components.rs [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+/// Default log level for SSE Events.
+// NOTE: Made this a constant. Debug level seems to be pretty intense. Can make this
+// configurable later if needed.
+const LOG_LEVEL: slog::Level = slog::Level::Info;
+
+/// The components required in the HTTP API task to receive logged events.
+#[derive(Clone)]
+pub struct SSELoggingComponents {
+    /// The channel to receive events from.
+    pub sender: Arc<AssertUnwindSafe<Sender<AsyncRecord>>>,
```
Comment (paulhauner) [inline common/logging/src/sse_logging_components.rs:19]: I'm not familiar with `UnwindSafe`, so I did some reading. I think you're using it right here, I'll describe my reasoning to show I actually thought about it.

The [`std:panic::UnwindSafe` help page](https://doc.rust-lang.org/std/panic/trait.UnwindSafe.html) has some info and also points to [an RFC](https://github.com/rust-lang/rfcs/blob/master/text/1236-stabilize-catch-panic.md). From here, I've learned that since `UnwindSafe` doesn't introduce memory-safety concerns, it is mostly a "speed bump"; something designed to make developers stop and think about what happens during a panic. The RFC goes on to state *"safe Rust code can largely ignore exception safety concerns."*

I understand that we can generally ignore exception safety because:

1. Memory safety is already guaranteed (assuming no `unsafe` code) and can't be violated be gregarious use of `UnwindSafe`. Therefore, unwind safety is mostly concerned about *logical* data structure integrity.
1. Catching panics is difficult/rare in Rust, so we're unlikely to actually observe any logical data structure issues created by a safe unwind.

So, given that we don't need to worry about memory-safety with `UnwindSafe`, I think the remaining questions is:

*If `tokio::sync::broadcast::Sender` panics and leaves itself in an inconsistent state, is `slog` going to observe this?*

My understanding is:

 *No, `slog` won't care if the `Sender` is in an invalid state.*
 
The only thing that `slog` can do is send things via `Drain::log` and we're already handling the `Result` from `sender.send`. Therefore, the `Sender` can panic as weirdly as it wants and `slog` will be none the wiser.

File: common/logging/src/sse_logging_components.rs [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+    }
+}
+
+impl Drain for SSELoggingComponents {
+    type Ok = ();
+    type Err = &'static str;
+
+    fn log(&self, record: &Record, logger_values: &OwnedKVList) -> Result<Self::Ok, Self::Err> {
+        if record.level().is_at_least(LOG_LEVEL) {
+            // There are subscribers, attempt to send the logs
```
Comment (paulhauner) [inline common/logging/src/sse_logging_components.rs]:
```suggestion
            // Attempt to send the logs.
```

I assume this is from a previous version where you were checking for subscribers first?

File: lighthouse/src/main.rs [hunk 5]
```diff
@@ -483,6 +483,16 @@ fn run<E: EthSpec>(
         };
     }
 
+    let sse_logging = {
+        if let Some(bn_matches) = matches.subcommand_matches("beacon_node") {
+            bn_matches.is_present("http") || bn_matches.is_present("gui")
```
Comment (paulhauner) [inline lighthouse/src/main.rs]:
```suggestion
            bn_matches.is_present("gui")
```

~~The `--staking` flag also turns on the HTTP API.~~

Alternatively, what are your thoughts on only enabling this when the `--gui` flag is present? It would be make the roll-out softer, at the moment all stakers will be running elements of the async logging system in v4.2.0. I think its OK as-is, just running the idea past you.

File: beacon_node/http_api/src/lib.rs [hunk 6]
```diff
@@ -3721,6 +3726,44 @@ pub fn serve<T: BeaconChainTypes>(
...
+                                    match data.to_json_string() {
+                                        // Send the json as a Server Side Event
+                                        Ok(json) => Ok(Event::default().data(json)),
+                                        Err(e) => Err(warp_utils::reject::server_sent_event_error(
+                                            format!("Unable to serialize to JSON {}", e),
+                                        )),
+                                    }
+                                }
+                                Err(e) => Err(warp_utils::reject::server_sent_event_error(
+                                    format!("Unable to serialize to JSON {}", e),
```
Comment (paulhauner) [inline beacon_node/http_api/src/lib.rs]:
```suggestion
                                    format!("Unable to receive event {}", e),
```

It might help with troubleshooting; there's another error case with the same message.

File: validator_client/src/http_api/mod.rs [hunk 7]
```diff
@@ -1021,6 +1029,49 @@ pub fn serve<T: 'static + SlotClock + Clone, E: EthSpec>(
...
+                                                e
+                                            ))
+                                        }),
+                                        Err(e) => Err(warp_utils::reject::server_sent_event_error(
+                                            format!("Unable to serialize to JSON {}", e),
+                                        )),
+                                    }
+                                }
+                                Err(e) => Err(warp_utils::reject::server_sent_event_error(
+                                    format!("Unable to serialize to JSON {}", e),
```
Comment (paulhauner) [inline validator_client/src/http_api/mod.rs]:
```suggestion
                                    format!("Unable to receive event {}", e),
```

As above.

File: common/logging/src/async_record.rs [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+}
+
+impl AsyncRecord {
+    /// Serializes a `Record` and an `OwnedKVList`.
+    pub fn from(record: &Record, logger_values: &OwnedKVList) -> Self {
+        let mut ser = ToSendSerializer::new();
+        record
+            .kv()
+            .serialize(record, &mut ser)
+            .expect("`ToSendSerializer` can't fail");
```
Comment (paulhauner) [inline common/logging/src/async_record.rs:37]: I don't know enough to know that this will never panic. If you're certain it won't panic, then it's fine by me.

A non-panic alternative would be to return an `Option` here and then just don't send a message down the channel in `impl Drain for SSELoggingComponents` if this returns `None`.

My concern here would be someone figuring out how to make some graffiti/peer-id/etc that can crash our logger.

File: common/logging/src/async_record.rs [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+        fn emit_i128(&mut self, key: Key, val: i128) -> slog::Result {
+            impl_m!(self, key, &val)
+        }
+    }
+    fn emit_str(&mut self, key: Key, val: &str) -> slog::Result {
+        impl_m!(self, key, &val)
+    }
+    fn emit_arguments(&mut self, key: Key, val: &fmt::Arguments) -> slog::Result {
+        TL_BUF.with(|buf| {
+            let mut buf = buf.borrow_mut();
```
Comment (paulhauner) [inline common/logging/src/async_record.rs:300]: I understand that we're using this thread-local `RefCell` to avoid allocating a new string for each log message. That seems like a nice thing.

We could use the non-panicking [`try_borrow_mut`](https://doc.rust-lang.org/std/cell/struct.RefCell.html#method.try_borrow_mut) variant here and if there's concurrent access to `buf` then we could return an `Err`.

I don't see how it's possible to get concurrent access to `buf` with this code, however I haven't used thread-locals and `RefCell`s much.

**Review by paulhauner — CHANGES_REQUESTED** [review]
Alrighty, here's my review! There's certainly some stuff going on here, I had to do some learning!

File: common/logging/src/sse_logging_components.rs (hunk 3 above)
Comment (AgeManning) [reply common/logging/src/sse_logging_components.rs:19]: Yeah. I went down this rabbit hole too. Had a bit of discussion with michael about it also. Came to the same conclusion

File: lighthouse/src/main.rs (hunk 5 above)
Comment (AgeManning) [reply lighthouse/src/main.rs]: Yeah I forgot the reason I did this. 

I think we enable this as minimally as we can. On first thought seems right to only enable on gui. I'll have a think and modify if needed.

File: common/logging/src/async_record.rs (hunk 8 above)
Comment (AgeManning) [reply common/logging/src/async_record.rs:37]: Yeah, so slog has this `Serializer` trait. The trait returns an `slog::Result`. 

If you look here: https://github.com/AgeManning/lighthouse/blob/sse-logs/common/logging/src/async_record.rs#LL68C1-L68C1

This implements every single method of the trait. Every singe function returns Ok(()), so its impossible to return an Err. The reason it has to return a Result is just to match the trait return type. 

The slog_async crate (that we use for all our logging) has the same logic: https://docs.rs/slog-async/latest/src/slog_async/lib.rs.html#480

tldr; we're either all good, or we're in real big trouble ;P

File: common/logging/src/async_record.rs (hunk 9 above)
Comment (AgeManning) [reply common/logging/src/async_record.rs:300]: This is borrowed from `slog_serde`. I assume their method has merit, happy to change tho. See here: https://docs.rs/slog-serde/1.0.0-alpha9/src/slog_serde/lib.rs.html#18

Comment (paulhauner) [reply common/logging/src/async_record.rs:300]: Seems fine to me, happy to leave it as-is.

File: common/logging/src/async_record.rs (hunk 8 above)
Comment (paulhauner) [reply common/logging/src/async_record.rs:37]: Fine by me!

File: lighthouse/src/main.rs (hunk 5 above)
Comment (paulhauner) [reply lighthouse/src/main.rs]: I just updated my suggestion to only enable it with the `--gui` flag. That should mean you can enact all of the suggestions in my review via the Github UI.

File: common/logging/src/sse_logging_components.rs (hunk 4 above)
Comment (AgeManning) [reply common/logging/src/sse_logging_components.rs]: It should just be `attempt to send the logs`. The send() itself checks if there are subscribers.

Comment (paulhauner) [reply common/logging/src/sse_logging_components.rs]: Suggestion updated.

**Review by paulhauner — APPROVED** [review]
Nice!

**Comment by michaelsproul** [comment]
bors retry

**Comment by paulhauner** [comment]
It seems like the `eth1-simulator-ubuntu` is consistently timing out: https://github.com/sigp/lighthouse/actions/runs/5003142484/jobs/8964003366

I haven't looked into it so I'm not sure why.

**Comment by AgeManning** [comment]
It seems to time out on other PRs also. I think its related to: https://github.com/sigp/lighthouse/pull/3555

**Comment by paulhauner** [comment]
`eth1-simulator-ubuntu` passed 🎉 

bors r+
